fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/AvailableFields.java

```diff
@@ -0,0 +1,211 @@
+            return null;
+        }
+
+        return builder;
+    }
+
+    @Nonnull
+    public static AvailableFields fromIndex(@Nonnull RecordType recordType,
+                                            @Nonnull Index index,
+                                            @Nonnull PlannableIndexTypes indexTypes,
+                                            @Nullable KeyExpression commonPrimaryKey) {
+        final KeyExpression rootExpression = index.getRootExpression();
+
+        final List<KeyExpression> keyFields = new ArrayList<>();
+        final List<KeyExpression> valueFields = new ArrayList<>();
+        if (indexTypes.getTextTypes().contains(index.getType())) {
+            // Full text index entries have all of their fields except the tokenized one.
+            keyFields.addAll(TextScanPlanner.getOtherFields(rootExpression));
+        } else if (indexTypes.getValueTypes().contains(index.getType()) ||
+                   indexTypes.getRankTypes().contains(index.getType())) {
+            keyFields.addAll(KeyExpression.getKeyFields(rootExpression));
+            valueFields.addAll(KeyExpression.getValueFields(rootExpression));
+        } else {
+            // Aggregate index
+            if (rootExpression instanceof GroupingKeyExpression) {
+                GroupingKeyExpression groupingKeyExpression = (GroupingKeyExpression) rootExpression;
+                keyFields.addAll(groupingKeyExpression.getGroupingSubKey().normalizeKeyForPositions());
+            }
+        }
+
+        // Like FDBRecordStoreBase.indexEntryKey(), but with key expressions instead of actual values.
+        final List<KeyExpression> primaryKeys = commonPrimaryKey == null
+                                                ? Collections.emptyList()
+                                                : commonPrimaryKey.normalizeKeyForPositions();
+        index.trimPrimaryKey(primaryKeys);
+        keyFields.addAll(primaryKeys);
+
+        Map<KeyExpression, FieldData> fields = new HashMap<>();
+        final IndexKeyValueToPartialRecord.Builder builder = IndexKeyValueToPartialRecord.newBuilder(recordType);
+        for (int i = 0; i < keyFields.size(); i++) {
+            KeyExpression keyField = keyFields.get(i);
+            FieldData fieldData = FieldData.of(IndexKeyValueToPartialRecord.TupleSource.KEY, i);
+            if (!keyField.createsDuplicates() && addCoveringField(keyField, fieldData, builder)) {
+                fields.put(keyField, fieldData);
+            }
+        }
+        for (int i = 0; i < valueFields.size(); i++) {
+            KeyExpression valueField = valueFields.get(i);
+            FieldData fieldData = FieldData.of(IndexKeyValueToPartialRecord.TupleSource.VALUE, i);
+            if (!valueField.createsDuplicates() && addCoveringField(valueField, fieldData, builder)) {
+                fields.put(valueField, fieldData);
+            }
+        }
+
+        if (!builder.isValid()) {
+            return NO_FIELDS;
```

Comment:
No, this can happen for fairly pedestrian reasons, such as having a repeated field. `isValid()` is a bit harsh sounding, but I didn't want to change the existing method name.